File: SeriesGuide/src/amazon/java/com/battlelancer/seriesguide/billing/amazon/AmazonBillingActivity.java
/*
 * Copyright 2014 Uwe Trottmann
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.battlelancer.seriesguide.billing.amazon;

import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;
import butterknife.Bind;
import butterknife.ButterKnife;
import com.amazon.device.iap.PurchasingService;
import com.amazon.device.iap.model.Product;
import com.amazon.device.iap.model.RequestId;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.ui.BaseActivity;
import com.battlelancer.seriesguide.util.Utils;
import timber.log.Timber;

public class AmazonBillingActivity extends BaseActivity {

    @Bind(R.id.progressBarAmazonBilling) View progressBar;
    @Bind(R.id.buttonAmazonBillingSubscribe) Button buttonSubscribe;
    @Bind(R.id.textViewAmazonBillingSubPrice) TextView textViewPriceSub;
    @Bind(R.id.buttonAmazonBillingGetPass) Button buttonGetPass;
    @Bind(R.id.textViewAmazonBillingPricePass) TextView textViewPricePass;
    @Bind(R.id.textViewAmazonBillingExisting) TextView textViewIsSupporter;
    @Bind(R.id.buttonAmazonBillingDismiss) Button buttonDismiss;
    @Bind(R.id.textViewAmazonBillingMoreInfo) View buttonMoreInfo;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_amazon_billing);
        setupActionBar();

        setupViews();

        AmazonIapManager.setup(this);
    }

    @Override
    protected void setupActionBar() {
        super.setupActionBar();
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
    }

    private void setupViews() {
        ButterKnife.bind(this);

        buttonSubscribe.setEnabled(false);
        buttonSubscribe.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                subscribe();
            }
        });

        buttonGetPass.setEnabled(false);
        buttonGetPass.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                purchasePass();
            }
        });

        buttonDismiss.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                dismiss();
            }
        });

        buttonMoreInfo.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Utils.launchWebsite(v.getContext(), getString(R.string.url_whypay),
                        "AmazonBillingActivity", "WhyPayWebsite");
            }
        });

        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    protected void onStart() {
        super.onStart();

        // no need to get product data every time we were hidden, so do it in onStart
        AmazonIapManager.get().requestProductData();
    }

    @Override
    protected void onResume() {
        super.onResume();

        AmazonIapManager.get().activate();
        AmazonIapManager.get().requestUserDataAndPurchaseUpdates();
    }

    @Override
    protected void onPause() {
        super.onPause();

        AmazonIapManager.get().deactivate();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            super.onBackPressed();
            return true;
        }
        return false;
    }

    private void subscribe() {
        final RequestId requestId = PurchasingService.purchase(AmazonSku.SERIESGUIDE_SUB.getSku());
        Timber.d("subscribe: requestId (" + requestId + ")");
    }

    private void purchasePass() {
        final RequestId requestId = PurchasingService.purchase(
                AmazonSku.SERIESGUIDE_PASS.getSku());
        Timber.d("purchasePass: requestId (" + requestId + ")");
    }

    private void dismiss() {
        finish();
    }

    public void onEventMainThread(AmazonIapManager.AmazonIapMessageEvent event) {
        Toast.makeText(this, event.messageResId, Toast.LENGTH_LONG).show();
    }

    public void onEventMainThread(AmazonIapManager.AmazonIapAvailabilityEvent event) {
        if (progressBar == null || buttonSubscribe == null || buttonGetPass == null
                || textViewIsSupporter == null) {
            return;
        }
        progressBar.setVisibility(View.GONE);

        // enable or disable purchase buttons based on what can be purchased
        buttonSubscribe.setEnabled(event.subscriptionAvailable && !event.userHasActivePurchase);
        buttonGetPass.setEnabled(event.passAvailable && !event.userHasActivePurchase);

        // status text
        if (!event.subscriptionAvailable && !event.passAvailable) {
            // neither purchase available, probably not signed in
            textViewIsSupporter.setText(R.string.subscription_not_signed_in);
        } else {
            // subscription or pass available
            // show message if either one is active
            textViewIsSupporter.setText(
                    event.userHasActivePurchase ? getString(R.string.upgrade_success) : null);
        }
    }

    public void onEventMainThread(AmazonIapManager.AmazonIapProductEvent event) {
        Product product = event.product;
        // display the actual price like "1.23 C"
        String price = product.getPrice();
        if (price == null) {
            price = "--";
        }
        if (AmazonSku.SERIESGUIDE_SUB.getSku().equals(product.getSku())) {
            if (textViewPriceSub != null) {
                textViewPriceSub.setText(
                        getString(R.string.billing_price_subscribe, price));
            }
            return;
        }
        if (AmazonSku.SERIESGUIDE_PASS.getSku().equals(product.getSku())) {
            if (textViewPricePass != null) {
                textViewPricePass.setText(
                        price + "\n" + getString(R.string.billing_price_pass));
            }
        }
    }
}

File: app/src/amazon/java/com/battlelancer/seriesguide/billing/amazon/AmazonBillingActivity.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2014-2025 Uwe Trottmann

package com.battlelancer.seriesguide.billing.amazon

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import com.amazon.device.iap.PurchasingService
import com.battlelancer.seriesguide.R
import com.battlelancer.seriesguide.billing.amazon.AmazonIapManager.AmazonIapAvailabilityEvent
import com.battlelancer.seriesguide.billing.amazon.AmazonIapManager.AmazonIapMessageEvent
import com.battlelancer.seriesguide.billing.amazon.AmazonIapManager.AmazonIapProductEvent
import com.battlelancer.seriesguide.databinding.ActivityAmazonBillingBinding
import com.battlelancer.seriesguide.ui.BaseActivity
import com.battlelancer.seriesguide.util.ThemeUtils
import com.battlelancer.seriesguide.util.ViewTools
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import timber.log.Timber

/**
 * Offers a single subscription using the Amazon in-app purchasing library.
 *
 * To debug: install on device with Amazon App Store and download App Tester app, put
 * `amazon.sdktester.json` onto `sdcard`.
 * Run command `adb shell setprop debug.amazon.sandboxmode debug` and launch app tester, then app.
 * Debug version of app works.
 *
 * To disable sandbox mode run `adb shell setprop debug.amazon.sandboxmode none`.
 *
 * 2022-05-19: subscription not recognized because sku in receipt is null, entitlement works.
 * Confirmed that subscription works and is recognized with production version.
 *
 * https://developer.amazon.com/de/docs/in-app-purchasing/iap-app-tester-user-guide.html
 */
class AmazonBillingActivity : BaseActivity() {

    private lateinit var binding: ActivityAmazonBillingBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityAmazonBillingBinding.inflate(layoutInflater)
        setContentView(binding.root)
        ThemeUtils.configureForEdgeToEdge(binding.root)
        setupActionBar()
        setupViews()
        AmazonHelper.create(this)
        AmazonHelper.iapManager.register()
    }

    override fun setupActionBar() {
        super.setupActionBar()
        val actionBar = supportActionBar
        if (actionBar != null) {
            actionBar.setHomeAsUpIndicator(R.drawable.ic_clear_24dp)
            actionBar.setDisplayHomeAsUpEnabled(true)
        }
    }

    private fun setupViews() {
        ThemeUtils.applyBottomPaddingForNavigationBar(binding.scrollViewAmazonBilling)

        binding.buttonAmazonBillingSubscribe.isEnabled = false
        binding.buttonAmazonBillingSubscribe.setOnClickListener { subscribe() }

        ViewTools.openUriOnClick(
            binding.textViewAmazonBillingMoreInfo,
            getString(R.string.url_billing_info_and_help)
        )
        binding.progressBarAmazonBilling.visibility = View.VISIBLE
    }

    override fun onStart() {
        super.onStart()

        // no need to get product data every time we were hidden, so do it in onStart
        AmazonHelper.iapManager.requestProductData()
    }

    override fun onResume() {
        super.onResume()
        AmazonHelper.iapManager.activate()
        AmazonHelper.iapManager.requestUserDataAndPurchaseUpdates()
    }

    override fun onPause() {
        super.onPause()
        AmazonHelper.iapManager.deactivate()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressedDispatcher.onBackPressed()
            return true
        }
        return false
    }

    private fun subscribe() {
        val requestId = PurchasingService.purchase(
            AmazonSku.SERIESGUIDE_SUB_YEARLY.sku
        )
        Timber.d("subscribe: requestId (%s)", requestId)
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEventMainThread(event: AmazonIapMessageEvent) {
        Toast.makeText(this, event.messageResId, Toast.LENGTH_LONG).show()
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEventMainThread(event: AmazonIapAvailabilityEvent) {
        binding.progressBarAmazonBilling.visibility = View.GONE

        // enable or disable purchase buttons based on what can be purchased
        binding.buttonAmazonBillingSubscribe.isEnabled =
            event.subscriptionAvailable && !event.userHasActivePurchase

        // status text
        if (!event.subscriptionAvailable) {
            // neither purchase available, probably not signed in
            binding.textViewAmazonBillingExisting.setText(R.string.subscription_not_signed_in)
        } else {
            // subscription available
            binding.textViewAmazonBillingExisting.text =
                if (event.userHasActivePurchase) getString(R.string.upgrade_success) else null
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEventMainThread(event: AmazonIapProductEvent) {
        val product = event.product
        // display the actual price like "1.23 C"
        var price = product.price
        if (price == null) {
            price = "--"
        }
        if (AmazonSku.SERIESGUIDE_SUB_YEARLY.sku == product.sku) {
            val priceString = getString(R.string.billing_duration_format, price)
            val trialInfo = getString(R.string.billing_sub_description)
            val finalPriceString = "$priceString\n$trialInfo"
            binding.textViewAmazonBillingSubPrice.text = finalPriceString
        }
    }
}